### DISCUSSION TITLE
Is there a way to build for local testing vs production deployment?

### DISCUSSION THREAD

**Discussion by Jefficus** [getnikola/nikola#3670] [Q&A] [answered]
Currently, if I want to do some local work (tweaking the theme, experimenting with a plugin, etc.) I have to change a bunch of conf.py settings and then remember to change them back before the next deploy. I was hoping to find a way to override specific settings with environment vars or at least be able to point to a different conf.py on the command line, but I'm not finding anything like that.

How are other people managing their developer vs production contexts?

**Comment by Kwpolska** [comment]
I would recommend keeping your site in a Git repository. That way, you can easily experiment and then revert the commit or undo uncommitted changes.

**Reply by Jefficus** [reply]
I _do_ have my site in a git repo. (I'm hosting it on GitHub, so it's pretty much required. :-)

It just seems like huge overkill to branch the repo and modify the conf.py, just to change a few color settings. I'd like to be able to make the changes, build locally, and look at it in the browser on my local machine a few times without all the overhead. If I don't like the changes, I could just checkout-stomp the local files, and if I do like them, I could commit and deploy. Easy peasy.

To do that, we would need to be able to do something like: `SITE_URL=localhost:8000 nikola build` and check our work before pushing it out into the world. (And I would argue that this kind of workflow would be much easier to explain to newbies than how to work with `git branch`.)

**Reply by Kwpolska** [reply]
You can already do that? As long as your output folder is not the folder you’re serving from (and our docs don’t ever suggest this route), you can run `nikola build` and `nikola serve` (or `nikola auto`) and just see things locally. And then you can do a `git restore` and `nikola build` again before you deploy.

**Reply by Jefficus** [reply]
Yes, that's the way I was doing it at first, but it turns out that doesn't work for things like RSS index pages, because they embed absolute URLs based on the SITE_URL setting. So that led me down the path of trying to have local and production versions of the conf.py. (Which I can now do, thanks to the `--conf` option that @felixfontein pointed out.)

Thanks for taking the time to help me muddle through it.

**Comment by felixfontein** [comment, marked as answer]
You could have two config files, one for testing and one for production, and select the one to use with the `--conf=` parameter.

**Reply by Jefficus** [reply]
OMG! That's _exactly_ what I've been looking for, but I haven't seen that option documented _anywhere_! Thank you.

Out of curiosity did I miss that param in the documentation somewhere? `nikola build --help` and `man nikola` don't seem to mention it.

**Reply by felixfontein** [reply]
It is mentioned in the handbook: https://getnikola.com/handbook.html#configuration

BTW, in the config file you probably want to set `CACHE_FOLDER` and `DOIT_CONFIG.dep_file` to something different (for at least one of the two configs), like this:
```py
CACHE_FOLDER = 'cache-testing'

DOIT_CONFIG = {
    'dep_file': '.doit-testing.db'
}
```

**Comment by Jefficus** [comment]
D'oh! I've been using the handbook extensively but I assumed `nikola
--help` was complete so never thought to look in the book for more options.
My bad.

And now that I read up on it, I see that we can use minimal alternate
configs that only override specific variables from our normal conf, which
completely solves my hope to do that with environment variables.

I love it when I discover that a new tool is even more robust than I
thought. Thanks again.
